platform/mbed_error.h

```diff
@@ -66,12 +834,218 @@
+ * @param  error_msg        The error message to be printed out to STDIO/Serial.
+ * @param  error_value      Value associated with the error status. This would depend on error code/error scenario.
+ * @param  filename         Name of the source file originating the error( Most callers can pass __FILE__ here ).
+ * @param  line_number      The line number of the source file originating the error( Most callers can pass __LINE__ here ) .
+ * @return                  0 or ERROR_SUCCESS.
+ *                          ERROR_INVALID_ARGUMENT if called with invalid error status/codes 
+ *
+ * @code
+ * 
+ * set_error( ERROR_OUT_OF_MEMORY, "Out of memory error", 0, __FILE__, __LINE__ )
+ *
+ * @endcode
+ *
+ * @note See SET_ERROR/SET_ERROR_FATAL macros which provides a wrapper on this API
+ */
+MbedErrorStatus set_error(MbedErrorStatus error_status, const char *error_msg, unsigned int error_value, const char *filename, int line_number);
+
+/**
+ * Returns the first system error reported.
+ * @return                  MbedErrorStatus code logged for the first error or ERROR_SUCCESS if no errors are logged.
+ *
+ */
+MbedErrorStatus get_first_error(void);
+
+/**
+ * Returns the most recent system error reported.
+ * @return                  MbedErrorStatus code logged for the last error or ERROR_SUCCESS if no errors are logged.
+ *
+ */
+MbedErrorStatus get_last_error(void);
+
+/**
+ * Returns the number of system errors reported after boot.
+ * @return                  int Number of errors reported.
+ *
+ */
+int get_error_count(void);
+
+/**
+ * Call this function to set a fatal system error and halt the system. This function will log the fatal error with the context info and prints the error report.
+ * 
+ * @param  error_status     MbedErrorStatus status to be set(See MbedErrorStatus enum above for available error status values).
+ * @param  error_msg        The error message to be printed out to STDIO/Serial.
+ * @param  error_value      Value associated with the error status. This would depend on error code/error scenario.
+ * @param  filename         Name of the source file originating the error( Most callers can pass __FILE__ here ).
+ * @param  line_number      The line number of the source file originating the error( Most callers can pass __LINE__ here ) .
+ * @return                  0 or ERROR_SUCCESS.
+ *                          ERROR_INVALID_ARGUMENT if called with invalid error status/codes 
+ *
+ * @code
+ * 
+ * set_error_fatal( ERROR_PROHIBITED_OPERATION, "Prohibited operation tried", 0, __FILE__, __LINE__ )
+ *
+ * @endcode
+ *
+ * @note See SET_ERROR/SET_ERROR_FATAL macros which provides a wrapper on this API
+ */
+MbedErrorStatus set_error_fatal(MbedErrorStatus error_status, const char *error_msg, unsigned int error_value, const char *filename, int line_number);
+
+/**
+ * Registers an application defined error callback with the error handling system. 
```

Comment:
Well, are you saying we need another hook for Hardfault?. I think it will still be the same, the only thing which is different is the kind of info. Should we still need another hook?

Reply:
One should be enough and it's simpler, as long as it's called on error and not warning aka in a situation like hard fault. It should be the last thing to happen, so user can reboot the platform or do whatever they wish.

Reply:
Ok, Ill look into how we can change the error handling.